### ISSUE TITLE
to_param method being deprecated gives warnings when params passed in a path helper

### ISSUE THREAD

**Issue by tomdracz** [rails/rails#26958]
### Steps to reproduce

Create params object and pass it to a path helper

```ruby
    params = ActionController::Parameters.new
    params[:test] = {}
    params[:test][:truthy] = "1"
    params[:test][:falsey] = "0"

example_path(params: { params[:test] })
```

### Expected behavior
No deprecation messages

### Actual behavior
Deprecation warning as to_param method is a Hash thing

```bash
Method to_param is deprecated and will be removed in Rails 5.1, as `ActionController::Parameters` no longer inherits from hash. Using this deprecated behavior exposes potential security problems. If you continue to use this method you may be creating a security vulnerability in your app that can be exploited. Instead, consider using one of these documented methods which are not deprecated: http://api.rubyonrails.org/v5.0.0.1/classes/ActionController/Parameters.html
```

### Some background

We're using Ransack gem to handle search in our data tables. The query string ends up being params[:q] With this being now ActionController::Parameters, instead of hash it results in deprecation warning when passing params[:q] to path helper as an argument (as it's trying to call to_param on params[:q]) i.e.:

```ruby
search_path(format: "csv", params: { data_table_id: params[:data_table_id], q: params[:q] })
```

Gives a deprecation warning and will break in Rails 5.1

### System configuration
**Rails version**: From 5-0-stable branch

**Ruby version**: 2.3.1p112

**Comment by jonasporto** [comment]
I didn't receive any deprecation warnings while using ```Rails 5.0.0.1```

I tried with given steps adding a key ```:test``` to avoid syntax errors when try to evaluate hash  ```example_path(params: { test: params[:test] })```

**Comment by tomdracz** [comment]
Rails 5.0.0.1 suffers from a different issue. Looks like something similar to https://github.com/rails/rails/issues/26802 where params object is not evaluated and ends up as ActionController::Params

I'm using '5-0-stable' branch. Below is the app do demonstrate the behaviour
https://github.com/tomdracz/params_path_test

Start the app, go to localhost:3000 or localhost:3000/pages There's a single link there like:
```ruby
<%= link_to "Test Path", pages_path(params: { test: params[:test] }) %>
```
params are set in the controller:
```ruby
class PagesController < ApplicationController
  def index
    params[:test] = {}
    params[:test][:truthy] = "1"
    params[:test][:falsey] = "1"
  end
end
```

Hiting the pages above gives a deprecation error

Actually is that issue is a duplicate and fixed in https://github.com/rails/rails/pull/26810 ? Looks like it might be

**Comment by jonasporto** [comment]
I agree really is a new problem in rails 5.0.0.1, is needed opening a new issue.

I did see the awkward behaviour in app example that you provide.

I think that is something related with when the params is binded to view context and try to use `#to_hash` in a `ActionController::Parameter` instance

we can simulate the behavior in `rails console`
```
2.3.1 :003 > ActionController::Parameters.new.to_hash
DEPRECATION WARNING: Method to_hash is deprecated and will be removed in Rails 5.1, as `ActionController::Parameters` no longer inherits from hash. Using this deprecated behavior exposes potential security problems. If you continue to use this method you may be creating a security vulnerability in your app that can be exploited. Instead, consider using one of these documented methods which are not deprecated: http://api.rubyonrails.org/v5.0.0.1/classes/ActionController/Parameters.html 
```

**Comment by y-yagi** [comment]
I think this fixed by #28734.

**Comment by tomdracz** [comment]
Seems like it. Closing tis now then
